### ISSUE TITLE
Clean up linking pass

### ISSUE THREAD

**Issue by onbjerg** [foundry-rs/foundry#975]
The linking function in `foundry-utils` is very hard to parse ([here](https://github.com/gakonst/foundry/blob/db909f286715ad5e4e045aae99fd6388a4a4b2b3/utils/src/lib.rs#L903-L992) and [here](https://github.com/gakonst/foundry/blob/db909f286715ad5e4e045aae99fd6388a4a4b2b3/utils/src/lib.rs#L89-L150)), we should investigate if there is already better parts in `ethers-solc` or make some

**Comment by onbjerg** [comment]
Here is a draft I've thought about.

We introduce two new traits in `ethers-solc`. The first one is `Linkable`:

```rs
/// A linkable object
trait Linkable {
    /// Tries to link the bytecode object with the `file` and `library` name.
    /// Replaces all library placeholders with the given address.
    ///
    /// Returns true if the bytecode object is fully linked, false otherwise
    /// This is a noop if the bytecode object is already fully linked.
   pub fn link(
        &mut self,
        file: impl AsRef<str>,
        library: impl AsRef<str>,
        address: Address,
    ) -> bool;
    /// Links the bytecode object with all provided `(file, lib, addr)`
    pub fn link_all<I, S, T>(&mut self, libs: I) -> bool
    where
        I: IntoIterator<Item = (S, T, Address)>,
        S: AsRef<str>,
        T: AsRef<str>;

    /// Same as `Bytecode::link` but with fully qualified name (`file.sol:Math`)
    pub fn link_fully_qualified(&mut self, name: impl AsRef<str>, addr: Address) -> bool;
    /// Links the bytecode object with all provided `(fully_qualified, addr)`
    pub fn link_all_fully_qualified<I, S>(&mut self, libs: I) -> bool
    where
        I: IntoIterator<Item = (S, Address)>,
        S: AsRef<str>;

    /// Whether the object is fully linked or not.
    pub fn is_linked(&self) -> bool;
    /// Whether the object is fully linked or not.
    pub fn is_unlinked(&self) -> bool;
}
```

`Linkable` is implemented for:

- `Artifact`: A proxy for linking the artifact's bytecodes
- `Bytecode`: Links the bytecode directly
- `CompactBytecode`: Links the bytecode directly
- `BytecodeObject`: Links the bytecode if it's unlinked
- `Contract`: A proxy for linking the contract's bytecodes
- `CompactContract`: A proxy for linking the contract's bytecodes
- `ProjectCompileOutput`: A proxy for linking many artifacts at the same time
- `Artifacts`: A proxy for linking many artifacts at the same time

These work the same as what we have now for `Bytecode`.

Next, we have the `DynamicallyLinkable` trait, and some helper types:

```rs
/// A map of library bytecode to the address it was deployed to
type Dependencies = Vec<(Address, Bytecode)>;

/// A map of artifact IDs to their [Dependencies].
type LinkerOutput = BTreeMap<ArtifactId, Dependencies>;

/// A linker function that is provided an [Artifact], a library's [Artifact],
/// the `index` of the library denoting the deployment order
/// for the library in respect to the provided contract.
///
/// The function should return the address at which the library is deployed.
type LinkerFn = FnMut(
    contract: (&ArtifactId, &Artifact),
    library: (&ArtifactId, &Artifact),
    index: usize
) -> Option<Address>;

trait DynamicallyLinkable {
    /// Links all contracts using the provided [LinkerFn].
    pub fn link_all<F>(&mut self, f: F) -> LinkerOutput;
}
```

This trait is implemented *only* for objects that have the required info, i.e.

- `Artifacts` (possibly?)
- `ProjectCompileOutput`

Calling `link_all` *should* try to link all contracts and their dependencies, taking nested libraries into account. The output of the function, `LinkerOutput`, should map every contract by `ArtifactId` to *every dependency* that contract and *its dependencies* would need, in order.

For example:

- Contract
  - Library
    - SecondLibrary
  - ThirdLibrary
- SecondContract
  - Library
    - SecondLibrary
- ThirdContract
  - ThirdLibrary
  - FourthLibrary
    - Library
      - SecondLibrary

Would call the linker function in this order (approximately)

- Contract
  - SecondLibrary, 0
  - Library, 1
  - ThirdLibrary, 2
- SecondContract
  - SecondLibrary, 0
  - Library, 1
- ThirdContract 
  - ThirdLibrary, 0
  - SecondLibrary, 1
  - Library, 2
  - FourthLibrary, 3

(*The format is (library, index)*)

The ordering matters, since this is the order Forge will deploy them in.

For `forge test`, the linking flow could then be something like:

```rs
/// Precompute all the library addresses
let mut linker_output = output.link_all_dynamic(|_, (_, library), index| {
    library
        .get_bytecode_bytes()
        .map(|bytes| ethers_core::utils::get_contract_address(sender, index + bytes.len()))
});

/// All of these should now be linked - if they're not, we get an error
let contracts: BTreeMap<ArtifactId, CompactContractSome> = output
  .with_stripped_file_prefixes(root)
  .into_artifacts()
  .map(|(id, artifact)| {
    Ok((id, artifact.into_compact_contract().try_into()?))
  })
  .collect()?;

/// We filter and clone the contracts that we should deploy (tests)
/// `contracts` now represents `known_contracts`
let tests: BTreeMap<ArtifactId, (CompactContractSome, LinkerOutput)> = self.contracts.iter()
  .filter(|(_, contract)| {
    contract
      .abi
      .constructor
      .map_or(true, |c| c.inputs.is_empty())
      && abi
        .functions()
        .any(|func| func.name.starts_with("test"))
  })
  .map(|(id, contract)| {
    Ok((id.clone(), (contract.clone(), linker_output.remove(id)?)))
  })
  .collect()?;
```

In the runner we would then just deploy the `Dependencies` of each contract in order.

**Comment by mattsse** [comment]
that looks great!
we definitely need this also for the compiler abstraction anyway

**Comment by onbjerg** [comment]
@meetmangukiya if you have some time on your hands feel free to take a stab at this (most of it is in ethers-rs) 😊 Let me know what you think

**Comment by meetmangukiya** [comment]
yep, I will take a look at this and the other one you mentioned, thanks

**Comment by mattsse** [comment]
@joshieDo fyi I had to refactor some related code here #1715 not sure if all if it will be merged with #1715 but the general idea to clean up `link_with...` function is to refactor how the Contract runners are constructed and add more helper types like this

https://github.com/foundry-rs/foundry/pull/1715/files#diff-32873586f91a77d6d9ece12276b6cad64781b711062bf77834623ff6379f251fR7-R24

and eventually, move the utils link functions into the `MultiContractRunner` which basically represents the whole project and knows all contracts

**Comment by gakonst** [comment]
@mattsse see the approach in https://github.com/gakonst/ethers-rs/pull/1293 for the current candidate refactor for how to do this, based on bjerg's design

**Comment by zerosnacks** [comment]
@klkvr can I mark this ticket as resolved given all the improvements around linking in https://github.com/foundry-rs/compilers/tree/main ?

**Comment by klkvr** [comment]
> @klkvr can I mark this ticket as resolved given all the improvements around linking in https://github.com/foundry-rs/compilers/tree/main ?

yep, I think we can close this
